rust/src/lib.rs

```diff
@@ -18,18 +19,20 @@ where
 }
 
 /// An abstract data type which maintains a fifo-ordered sliding window.
-pub trait FifoWindow<Value, BinOp>: Clone
+pub trait FifoWindow<BinOp>: Clone
 where
-    BinOp: Operator,
+    BinOp: AggregateOperator,
 {
     /// Returns an empty window.
     fn new() -> Self;
+    /// Return the experimental name for the algorithm.
+    fn name() -> &'static str;
     /// Inserts a value at the back of the window.
-    fn push(&mut self, v: Value);
+    fn push(&mut self, v: BinOp::In);
     /// Removes a value at the front of the window (if any).
-    fn pop(&mut self) -> Option<Value>;
+    fn pop(&mut self) -> Option<BinOp::Out>;
```

Comment:
I think pop should not return anything. I can see from the implementation of some of the aggregation algorithms that it lowers the oldest item and returns that. There are two problems with that. First, it is wasteful to do the lowering if that return value gets discarded by the caller. Second, and more importantly, not all aggregation algorithms hold on to that oldest item, so it is unavailable to be lowered and returned. Specifically, Two-Stacks Lite and DABA Lite don't keep around singleton partials for the front portion of the queue, since that portion is optimized for eviction, where it makes more sense to keep around partials for larger sublists.

Reply:
My intention with this PR was not to fundamentally change the `FifoWindow` interface Klas previously created, so I think that should be discussed in a new issue and changed with a different PR. I'm going to create an issue to discuss this point.